src/components/FileUploaderV2/FileUploaderV2.js

```diff
@@ -0,0 +1,285 @@
+          multiple={multiple}
+          accept={accept}
+          name={name}
+          onChange={onChange}
+          onClick={evt => {
+            evt.target.value = null;
+          }}
+        />
+      </div>
+    );
+  }
+}
+
+export function FilenameV2({
+  iconDescription,
+  onKeyDown = () => {},
+  status = 'uploading',
+  style = {},
+  tabIndex = 0,
+  ...other
+}) {
+  switch (status) {
+    case 'uploading':
+      return (
+        <div
+          className="bx--loading"
+          style={{ ...style, width: '1rem', height: '1rem' }}
+          tabIndex={tabIndex}
+          onKeyDown={onKeyDown}
+          role="button"
+          {...other}>
+          <svg className="bx--loading__svg" viewBox="-42 -42 84 84">
+            <circle cx="0" cy="0" r="37.5" />
+          </svg>
+        </div>
+      );
+    case 'edit':
+      return (
+        <Icon
+          description={iconDescription}
+          className="bx--file-close"
+          icon={iconCloseSolid}
+          style={style}
+          tabIndex={tabIndex}
+          onKeyDown={onKeyDown}
+          role="button"
+          {...other}
+        />
+      );
+    case 'complete':
+      return (
+        <Icon
+          description={iconDescription}
+          className="bx--file-complete"
+          icon={iconCheckmarkSolid}
+          style={style}
+          tabIndex={tabIndex}
+          onKeyDown={onKeyDown}
+          role="button"
+          {...other}
+        />
+      );
+    default:
+      return null;
+  }
+}
+FilenameV2.propTypes = {
+  onKeyDown: PropTypes.func,
+  style: PropTypes.object,
+  status: PropTypes.oneOf(['edit', 'complete', 'uploading']),
+  tabIndex: PropTypes.number,
+};
+
+export default class FileUploaderV2 extends Component {
+  static propTypes = {
+    iconDescription: PropTypes.string,
+    buttonLabel: PropTypes.string,
+    buttonKind: ButtonTypes.buttonKind,
+    labelDescription: PropTypes.string,
+    labelTitle: PropTypes.string,
+    multiple: PropTypes.bool,
+    name: PropTypes.string,
+    onClick: PropTypes.func,
+    onChange: PropTypes.func,
+    className: PropTypes.string,
+    accept: PropTypes.arrayOf(PropTypes.string),
+    action: PropTypes.string,
+    upload: PropTypes.func,
+  };
+
+  static defaultProps = {
+    iconDescription: 'Provide icon description',
+    buttonLabel: 'Upload',
+    buttonKind: 'primary',
+    multiple: false,
+    onClick: () => {},
+    onChange: () => {},
+    accept: [],
+    upload: () => {},
+  };
+
+  state = {
+    files: [],
+  };
+
+  nodes = [];
+
+  handleChange = evt => {
+    evt.stopPropagation();
+    const files = [...this.state.files];
+    [...evt.target.files].forEach(file => {
+      const uuid = uid();
+      files.push({
+        uuid,
+        name: file.name,
+        size: file.size,
+        status: 'uploading',
+      });
+      const index = files.findIndex(file => file.uuid === uuid);
+      this.props
+        .upload({ file })
+        .then(() => {
```

Comment:
The code here seems to assume `props.upload()` returns a `Promise`, but the corresponding `defaultProps` (https://github.com/emyarod/carbon-components-react/blob/c4a18be/src/components/FileUploaderV2/FileUploaderV2.js#L171) doesn’t? I don’t know if it’s the direct cause or not, but I see `Uncaught TypeError: Cannot read property 'then' of undefined` when I tried to upload a file in `<FileUploaderV2>`’s story.

Reply:
I've made the default `upload` prop value a function that returns an empty Promise to remove the console warning.

Is there a scenario where the `FileUploader` would be used standalone without providing an `upload` function? It might be worth completely removing the standalone `FileUploader` story and just leaving the example application story, if that would be a better illustration of how to use the component

Reply:
Great point @emyarod! This leads me to a question - Could an application simply listen to `onChange`, upload the file, and set file uploading status via props once uploading is complete? In this way we can relegate the application-specific file uploading procedure to application code and let our file uploader component focus on visuals and user interactions.